<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Results" data-toc-modified-id="Results-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Results</a></span><ul class="toc-item"><li><span><a href="#Audience-Counts-by-Type-and-Vertical" data-toc-modified-id="Audience-Counts-by-Type-and-Vertical-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Audience Counts by Type and Vertical</a></span></li><li><span><a href="#AK-analysis:-Emails" data-toc-modified-id="AK-analysis:-Emails-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>AK analysis: Emails</a></span></li><li><span><a href="#AK-analysis:-CIDs" data-toc-modified-id="AK-analysis:-CIDs-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>AK analysis: CIDs</a></span></li><li><span><a href="#AK-analysis:-Addresses" data-toc-modified-id="AK-analysis:-Addresses-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>AK analysis: Addresses</a></span></li><li><span><a href="#AK-analysis:-Pinless-Audiences" data-toc-modified-id="AK-analysis:-Pinless-Audiences-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>AK analysis: Pinless Audiences</a></span></li></ul></li><li><span><a href="#Load-Libraries,-Modules" data-toc-modified-id="Load-Libraries,-Modules-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Load Libraries, Modules</a></span></li><li><span><a href="#Online-audiences-vs.-Offline-audiences" data-toc-modified-id="Online-audiences-vs.-Offline-audiences-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Online audiences vs. Offline audiences</a></span><ul class="toc-item"><li><span><a href="#results" data-toc-modified-id="results-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>results</a></span></li><li><span><a href="#data-wrangling" data-toc-modified-id="data-wrangling-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>data wrangling</a></span></li></ul></li><li><span><a href="#Data-store-audiences-vs.-non-data-store-audiences" data-toc-modified-id="Data-store-audiences-vs.-non-data-store-audiences-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Data store audiences vs. non-data store audiences</a></span><ul class="toc-item"><li><span><a href="#results" data-toc-modified-id="results-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>results</a></span></li><li><span><a href="#data-wrangling" data-toc-modified-id="data-wrangling-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>data wrangling</a></span></li></ul></li><li><span><a href="#Email-only-audiences" data-toc-modified-id="Email-only-audiences-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Email-only audiences</a></span><ul class="toc-item"><li><span><a href="#Results" data-toc-modified-id="Results-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Results</a></span><ul class="toc-item"><li><span><a href="#all-audiences" data-toc-modified-id="all-audiences-5.1.1"><span class="toc-item-num">5.1.1&nbsp;&nbsp;</span>all audiences</a></span></li><li><span><a href="#non-DS-audiences" data-toc-modified-id="non-DS-audiences-5.1.2"><span class="toc-item-num">5.1.2&nbsp;&nbsp;</span>non-DS audiences</a></span></li><li><span><a href="#DS-audiences" data-toc-modified-id="DS-audiences-5.1.3"><span class="toc-item-num">5.1.3&nbsp;&nbsp;</span>DS audiences</a></span></li></ul></li><li><span><a href="#Data-wrangling" data-toc-modified-id="Data-wrangling-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>Data wrangling</a></span></li></ul></li><li><span><a href="#CID-only-audiences" data-toc-modified-id="CID-only-audiences-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>CID-only audiences</a></span><ul class="toc-item"><li><span><a href="#Results" data-toc-modified-id="Results-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Results</a></span><ul class="toc-item"><li><span><a href="#all-audiences" data-toc-modified-id="all-audiences-6.1.1"><span class="toc-item-num">6.1.1&nbsp;&nbsp;</span>all audiences</a></span></li><li><span><a href="#non-DS-audiences" data-toc-modified-id="non-DS-audiences-6.1.2"><span class="toc-item-num">6.1.2&nbsp;&nbsp;</span>non-DS audiences</a></span></li><li><span><a href="#DS-audiences" data-toc-modified-id="DS-audiences-6.1.3"><span class="toc-item-num">6.1.3&nbsp;&nbsp;</span>DS audiences</a></span></li></ul></li></ul></li><li><span><a href="#Address-only-audiences" data-toc-modified-id="Address-only-audiences-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Address-only audiences</a></span><ul class="toc-item"><li><span><a href="#Results" data-toc-modified-id="Results-7.1"><span class="toc-item-num">7.1&nbsp;&nbsp;</span>Results</a></span><ul class="toc-item"><li><span><a href="#all-audiences" data-toc-modified-id="all-audiences-7.1.1"><span class="toc-item-num">7.1.1&nbsp;&nbsp;</span>all audiences</a></span></li><li><span><a href="#data-store" data-toc-modified-id="data-store-7.1.2"><span class="toc-item-num">7.1.2&nbsp;&nbsp;</span>data store</a></span></li></ul></li><li><span><a href="#Data-Wrangling" data-toc-modified-id="Data-Wrangling-7.2"><span class="toc-item-num">7.2&nbsp;&nbsp;</span>Data Wrangling</a></span></li></ul></li><li><span><a href="#PINless-imports-analysis" data-toc-modified-id="PINless-imports-analysis-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>PINless imports analysis</a></span><ul class="toc-item"><li><span><a href="#results" data-toc-modified-id="results-8.1"><span class="toc-item-num">8.1&nbsp;&nbsp;</span>results</a></span></li><li><span><a href="#appendix" data-toc-modified-id="appendix-8.2"><span class="toc-item-num">8.2&nbsp;&nbsp;</span>appendix</a></span></li></ul></li></ul></div>

# Results

## Audience Counts by Type and Vertical 
Active customers, enabled audiences, non-FBR audiences

In [117]:
import pandas as pd
audience_counts = {'Audience Type': ['Offline','Online', 'Total'], 'Data Store': [DMS_offline_audience_count, 
                                                                                  DMS_online_audience_count, 
                                                                                  DMS_offline_audience_count + DMS_online_audience_count],
        'All Audiences': [offline_audience_count,
                          online_audience_count, 
                          offline_audience_count + online_audience_count]
        }

aud_counts_table = pd.DataFrame(audience_counts, columns = ['Audience Type', 'Data Store', 'All Audiences'])
aud_counts_table

Audience Type  Data Store  All Audiences
0       Offline         565          15282
1        Online        1080           5057
2         Total        1645          20339

## AK analysis: Emails

Active customers, enabled audiences, non-FBR audiences where AK = email. Validated audiences have one AK in SQL statement. 

Raw email definition: plaintext string e.g. anthonyschultz@gmail.com


In [172]:
email_aks = {'Vertical': ['non-Data Store', 'Data Store', 'Total'],
             'Raw': [raw_email_aud_counts - raw_email_DS_aud_counts, 
                     raw_email_DS_aud_counts, 
                     raw_email_aud_counts], 
             'SHA256': [SHA256_aud_counts - SHA256_DS_aud_counts, 
                        SHA256_DS_aud_counts, 
                        SHA256_aud_counts],
             'MD5': [MD5_aud_counts - MD5_DS_aud_counts, 
                     MD5_DS_aud_counts, 
                     MD5_aud_counts], 
             'SHA1':[SHA1_aud_counts - SHA1_DS_aud_counts, 
                     SHA1_DS_aud_counts, 
                     SHA1_aud_counts],
             'All Email Audiences':[(raw_email_aud_counts - raw_email_DS_aud_counts + 
                                     (MD5_aud_counts - MD5_DS_aud_counts) +
                                    (SHA1_aud_counts - SHA1_DS_aud_counts) 
                                    ), (raw_email_DS_aud_counts + 
                                        SHA256_DS_aud_counts + 
                                        SHA1_DS_aud_counts)
                                        , raw_email_aud_counts +
                                    SHA256_aud_counts +
                                    MD5_aud_counts +
                                    SHA1_aud_counts]}

email_aks_table = pd.DataFrame(email_aks, columns = ['Vertical','Raw', 'SHA256', 'MD5', 'SHA1', 'All Email Audiences'])
email_aks_table.head()

Vertical   Raw  SHA256  MD5  SHA1  All Email Audiences
0  non-Data Store  5162      45   94    24                 5280
1      Data Store    82       6   15     8                   96
2           Total  5244      51  109    32                 5436

## AK analysis: CIDs

Active customers, enabled audiences, non-FBR audiences where AK = email. Validated audiences have one AK in SQL statement.

In [5]:
cid_aks = {'Vertical': ['non-Data Store', 'Data Store', 'Total'],
           'Offline': [4050, 242, 4292], 
           'Online': [986, 176, 1162],
           'All CID Audiences': [5036, 418, 5454]
          }
cid_aks_table = pd.DataFrame(cid_aks, columns = ['Vertical', 'Offline', 'Online', 'All CID Audiences'])
cid_aks_table

Vertical  Offline  Online  All CID Audiences
0  non-Data Store     4050     986               5036
1      Data Store      242     176                418
2           Total     4292    1162               5454

## AK analysis: Addresses

Active customers, enabled audiences, non-FBR audiences where AK = address fields and between 4-7 values.

In [6]:
address_aks = {'Vertical': ['non-Data Store', 'Data Store', 'Total'], 
               'Offline': [1043, 90, 1133],
               'Online': [12, 0, 12], 
               'All Address Audiences': [1055, 90, 1145]
              }
address_aks_table = pd.DataFrame(address_aks, columns = ['Vertical', 'Offline', 'Online', 'All Address Audiences'])
address_aks_table

Vertical  Offline  Online  All Address Audiences
0  non-Data Store     1043      12                   1055
1      Data Store       90       0                     90
2           Total     1133      12                   1145

## AK analysis: Pinless Audiences

Audiences with greater than 31 imports (50th percentile) and greater than 19% pinless imports (50th percentile). See section for methodology.

131 audiences fit this criteria

In [20]:
pinless_aks = {'Vertical': ['non-Data Store', 'Data Store', 'Total'], 
               'Email': [7, 0, 7], 
               'CID': [83, 36, 119], 
               'Address': [2, 0, 2]
                    }
pinless_aks_table = pd.DataFrame(pinless_aks, columns = ['Vertical', 'Email', 'CID', 'Address'])
pinless_aks_table

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3 entries, 0 to 2
Data columns (total 4 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   Vertical  3 non-null      object
 1   Email     3 non-null      int64 
 2   CID       3 non-null      int64 
 3   Address   3 non-null      int64 
dtypes: int64(3), object(1)
memory usage: 224.0+ bytes


# Load Libraries, Modules

In [12]:
# import basics

from google.cloud import bigquery
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
import pandas_gbq
import tqdm
pd.set_option('display.max_rows', 500)

# Online audiences vs. Offline audiences

In [63]:
## results

print("Total Enabled Audience:", offline_audience_count + online_audience_count)

# offline audiences: 15,707 
print("offline audience count:", offline_audience_count)

# online audiences: 5,117
print("online audience count:", online_audience_count)

Total Enabled Audience: 20339
offline audience count: 15282
online audience count: 5057


## data wrangling

query all audiences -> enabled, non-FBR

In [67]:
sql = """
SELECT *
FROM `corp-bi-us-prod.rldb.liveramp_customer_accounts` 
WHERE enabled = true AND customer_link_audience = false;
"""
lcas = pandas_gbq.read_gbq(sql, project_id='liveramp-ts-bigquery')

Downloading: 100%|██████████| 20937/20937 [00:04<00:00, 4607.40rows/s]

Time of data pull = 07/07/2020 12:08:00


query audience type summary -> enabled, non-FBR

In [66]:
sql = """
SELECT type, count(type) AS type_count
FROM `corp-bi-us-prod.rldb.liveramp_customer_accounts` 
WHERE enabled = true AND customer_link_audience = false
GROUP BY type
ORDER BY type_count DESC
"""

aud_sum = pandas_gbq.read_gbq(sql, project_id='liveramp-ts-bigquery')


Downloading: 100%|██████████| 9/9 [00:00<00:00, 41.98rows/s]

Time of data pull = 07/07/2020 12:08:00


In [68]:
# aud_sum = pd.read_csv('/Users/aschul/Documents/Documents - aschul1118mac/Code_Base/Investigations/20200417_AK_analysis/all_LCAs_summary.csv')
print(aud_sum.head(2))

offline_audience_count = aud_sum["type_count"][0]
online_audience_count = aud_sum["type_count"][1]
print(offline_audience_count)
print(online_audience_count)

                             type  type_count
0  OfflineLiverampCustomerAccount       15282
1   OnlineLiverampCustomerAccount        5057
15282
5057


# Data store audiences vs. non-data store audiences

In [111]:
## results
print("Data Provider customers:", dms_providers["customer_id"].count())

print("Offline DP audiences:", DMS_offline_audience_count)
print("Online DP audiences:", DMS_online_audience_count)
print("All DP audience:", DMS_offline_audience_count + DMS_online_audience_count)

Data Provider customers: 832
Offline DP audiences: 565
Online DP audiences: 1080
All DP audience: 1645


## data wrangling

query all data provider customers -> customer status = enabled

In [103]:
sql = """
SELECT dms_providers.customer_id
FROM `corp-bi-us-prod.rldb.dms_providers` AS dms_providers
  INNER JOIN `corp-bi-us-prod.rldb.customers` AS customers
    ON dms_providers.customer_id = customers.id
WHERE customers.status = 1
"""
dms_providers = pandas_gbq.read_gbq(sql, project_id='liveramp-ts-bigquery')


Downloading: 100%|██████████| 832/832 [00:00<00:00, 3296.85rows/s]


In [108]:
dms_providers["customer_id"].count()

832

query all data store audiences -> enabled, non-FBR audiences, customer status = enabled

In [90]:
sql = """
SELECT lr_customer_accounts.id AS audience_id, lr_customer_accounts.display_name AS audience_name, customers.name AS customer_name, customers.id AS customer_id, lr_customer_accounts.type
FROM `corp-bi-us-prod.rldb.dms_segments` AS dms_segments
INNER JOIN `corp-bi-us-prod.rldb.lrc_field_definitions` AS lrc_field_definitions
  ON dms_segments.lrc_field_definition_id = lrc_field_definitions.id
INNER JOIN `corp-bi-us-prod.rldb.liveramp_customer_accounts` AS lr_customer_accounts
  ON lrc_field_definitions.lr_customer_account_id = lr_customer_accounts.id
INNER JOIN `corp-bi-us-prod.rldb.customers` AS customers
  ON lr_customer_accounts.customer_id = customers.id
WHERE lr_customer_accounts.enabled = true AND lr_customer_accounts.customer_link_audience = false AND customers.status = 1
GROUP BY lr_customer_accounts.type, lr_customer_accounts.id, lr_customer_accounts.display_name, customers.name, customers.id
"""
dms_audiences = pandas_gbq.read_gbq(sql, project_id='liveramp-ts-bigquery')
dms_audiences

Downloading: 100%|██████████| 1645/1645 [00:00<00:00, 1805.04rows/s]


audience_id                                      audience_name  \
0           98099                          Epsilon - Evergreen Shelf   
1          156416   Audience Cloud - Facebook Distros - Old (PUBIDs)   
2          138149                               Kiip - AAID and IDFA   
3          116619                            Factual - Data [mobile]   
4           88826  Ameribase Digital / Lighthouse List - Onboardi...   
...           ...                                                ...   
1640       209786                                          Net Worth   
1641       442131  Bombora - DDS - BPA – BNP Media - 67120 - Cust...   
1642       384086  Swoop > Custom > Health Relevance Model > 35-6...   
1643       272749                                       P4Ad-Push-DE   
1644       464831         Chamberlain University - DataGuru Segments   

                            customer_name  customer_id  \
0                                 Epsilon       576169   
1                         Acxiom Internal       511196   
2                                    Kiip       537599   
3                                 Factual       533129   
4     Lighthouse List - Ameribase Digital       521086   
...                                   ...          ...   
1640                             Windfall       520876   
1641                              Bombora       514566   
1642                                Swoop       508536   
1643                              0ptimus       575329   
1644    LinkedIn - Chamberlain University       623671   

                                type  
0     OfflineLiverampCustomerAccount  
1      OnlineLiverampCustomerAccount  
2      OnlineLiverampCustomerAccount  
3      OnlineLiverampCustomerAccount  
4     OfflineLiverampCustomerAccount  
...                              ...  
1640  OfflineLiverampCustomerAccount  
1641   OnlineLiverampCustomerAccount  
1642   OnlineLiverampCustomerAccount  
1643  OfflineLiverampCustomerAccount  
1644   OnlineLiverampCustomerAccount  

[1645 rows x 5 columns]

In [97]:
DMS_audience_count = dms_audiences.groupby(['type']).count()
DMS_offline_audience_count = DMS_audience_count["audience_id"][0]
DMS_online_audience_count = DMS_audience_count["audience_id"][1]
print("offline data store audience count:", DMS_offline_audience_count)
print("online data store audience count:", DMS_online_audience_count)

offline data store audience count: 565
online data store audience count: 1080


# Email-only audiences

In [167]:
print("All audiences, count w/email as AK")
print("Raw email:", raw_email_aud_counts)
print("Hashed email:", MD5_aud_counts + SHA256_aud_counts + SHA1_aud_counts)
print("Hashed email breakdown \n MD5 email:", MD5_aud_counts)
print("SHA256 email:", SHA256_aud_counts)
print("SHA1 email:", SHA1_aud_counts, "\n")

print("Non-DS audiences, count w/email as AK")
print("Raw email:", raw_email_aud_counts - raw_email_DS_aud_counts)
print("MD5 email:", MD5_aud_counts - MD5_DS_aud_counts)
print("SHA256:", SHA256_aud_counts - SHA256_DS_aud_counts)
print("SHA1:", SHA1_aud_counts - SHA1_DS_aud_counts, "\n")

print("DS Audience counts w/emails as AK")
print("raw email audiences:", raw_email_DS_aud_counts)
print("MD5 email:", MD5_DS_aud_counts)
print("SHA256:", SHA256_DS_aud_counts)
print("SHA1:", SHA1_DS_aud_counts)

All audiences, count w/email as AK
Raw email: 5244
Hashed email: 192
Hashed email breakdown 
 MD5 email: 109
SHA256 email: 51
SHA1 email: 32 

Non-DS audiences, count w/email as AK
Raw email: 5162
MD5 email: 94
SHA256: 45
SHA1: 24 

DS Audience counts w/emails as AK
raw email audiences: 82
MD5 email: 15
SHA256: 6
SHA1: 8


## Data wrangling

query all audiences with AK -> enabled, non-FBR audiences, enabled customers

In [122]:
sql = """
SELECT
  audience_key_labels.id AS ak_id,
  liveramp_customer_accounts.id,
  audience_key_labels.key_index,
  audience_key_labels.label,
  liveramp_customer_accounts.customer_id,
  liveramp_customer_accounts.display_name,
  COUNT(audience_key_labels.id) AS count_AKs
FROM
  `corp-bi-us-prod.rldb.audience_key_labels` AS audience_key_labels
INNER JOIN
  `corp-bi-us-prod.rldb.liveramp_customer_accounts` AS liveramp_customer_accounts
ON
  audience_key_labels.liveramp_customer_account_id = liveramp_customer_accounts.id
INNER JOIN
  `corp-bi-us-prod.rldb.customers` AS customers
ON
  liveramp_customer_accounts.customer_id = customers.id
WHERE
  customers.status = 1
  AND liveramp_customer_accounts.enabled = TRUE
  AND liveramp_customer_accounts.customer_link_audience = FALSE
  AND lower(audience_key_labels.label) LIKE '%email%'
  AND audience_key_labels.key_index = 0
GROUP BY
  liveramp_customer_accounts.id,
  liveramp_customer_accounts.display_name,
  liveramp_customer_accounts.customer_id,
  audience_key_labels.key_index,
  ak_id,
  audience_key_labels.label
HAVING count_AKs = 1
"""
raw_emails_all_audiences = pandas_gbq.read_gbq(sql, project_id='liveramp-ts-bigquery')


Downloading: 100%|██████████| 5568/5568 [00:00<00:00, 8280.49rows/s]


In [123]:
# normalize email columns (raw, MD5, SHA1, SHA256)
raw_emails_all_audiences["label"].replace({"Email":"email", 
                             "Email1":"email",
                             "Email Address":"email",
                             "EmailAddress":"email", 
                             "Emails":"email",
                             "EMAIL":"email",
                             "Email_1":"email",
                             "EmailAddress":"email",
                             "email_address":"email",
                             "EMAIL_ADDRESS":"email",
                             "Email 1":"email",
                             "Email(s)":"email",
                             "email1":"email",
                             "EMAIL_ID":"email",
                             "email1_MD5":"MD5",
                             "emailmd5":"MD5",
                             "profiles.email_address":"email",
                             "EMAIL_ADDR":"email",
                             "SHA256EMAIL1":"email",
                             "email_addr":"email",
                             "MD5EMAIL":"MD5",
                             "EmailAddress":"email",
                             "EMAIL1":"email",
                             "cust_email_addr":"email",
                             "Email_Address":"email",
                             "md5_email":"MD5",
                             "Email_ID":"email",
                             "EMAIL_ADDR_TXT":"email",
                             "emailAddress":"email",
                             "sha1_email":"SHA1",
                             "MD5Email":"MD5",
                             "email_1":"email",
                             "LRI_G_SHA1EMAIL1":"SHA1",
                             "LRI_G_SHA256EMAIL1":"SHA256",
                             "loweremailmd5":"MD5",
                             "email_sha256":"SHA256",
                             "SHA256_EMAIL":"SHA256",
                             "MD5_Email":"MD5",
                             "LRI_DER_SHA256EMAIL1":"SHA256",
                             "email_SHA-1":"SHA1",
                             "sha256email":"SHA256",
                             "email address":"email",
                             "purch_email_addr":"email",
                             "HashedEmail":"hashed_email",
                             "P_EMAIL_ADDRESS":"email",
                             "Member Email Address":"email",
                             "Email Addresses":"email",
                             "email_md5":"MD5",
                             "emails":"email",
                             "Email Hash":"hashed_email",
                             "customer_email":"email",
                             "EMAIL_ADDRESS_":"email",
                             "SHA 256 Email":"SHA256",
                             "EMAILADDRESS":"email",
                             "LRI_DER_SHA1EMAIL1":"SHA1",
                             "email_hash":"hashed_email",
                             "LRI_G_MD5EMAIL1":"MD5",
                             "EMAIL_SHA256":"SHA256",
                             "PII_EMAIL_ADDRESS":"email",
                             "EMail":"email",
                             "Email address":"email",
                             "SHA1_Email":"SHA1",
                             "LRI_DER_MD5EMAIL1":"MD5",
                             "EMAIL_1":"email",
                             "sha256_email":"SHA256",
                             "SHA256EMAIL":"SHA256",
                             "SHA256Email":"email",
                             "emailaddress":"email",
                             "email_id":"email",
                             "SHA1EMAIL":"SHA1",
                             "hashed email":"hashed_email",
                             "SHA256HASHEDEMAIL":"SHA256",
                             "email2":"email",
                             "EMAIL_ADDRESS_1":"email",
                             "md5email":"MD5",
                             "hashedemail":"hashed_email",
                             "EMAIL_HASHED":"hashed_email",
                             "Broker Email":"email",
                             "HASH_MD5_LOWER_EMAIL":"MD5",
                             "profileEmail":"email",
                             "email_Address":"email",
                             "Hashed_Email":"hashed_email",
                             "PER_EMAIL":"email",
                             "Hashed_Email_Address":"email",
                             "Peloton Order Order Email":"email",
                             "EMAIL,FN,LN,,ZIP,CT,PHONE,":"email",
                             "Hashed_Email_ID_Hex":"hashed_email",
                             "#email_lc_md5":"MD5",
                             "MD5_EMAIL":"MD5",
                             "\EMAIL\"":"email",
                             "Customer Email":"email",
                             "CUSTOMERS_SalesOrder_BILLTOLookup::Email":"email",
                             "profile.email":"email",
                             "HASHED_EMAIL":"hashed_email",
                             "sha256email1":"SHA256",
                             "Hashed Email":"hashed_email",
                             "CUST_EMAIL":"email",
                             "SHA256 Email":"SHA256",
                             "email,fn,ln,st,zip,ct,phone,":"email",
                             "best_email":"email",
                             "LRI_G_EMAIL1":"email",
                             "SHA2_EMAIL":"SHA2",
                             "Md5 Email":"MD5",
                             "md5(email1)":"MD5",
                             "master_email":"email",
                             "HASHED_CUST_EMAIL":"hashed_email",
                             "REAL_EMAIL":"email",
                             "email_sha1":"SHA1",
                             "MD5 EMAIL":"MD5",
                             "Email/Email":"email",
                             "hash_email":"hashed_email",
                             "email md5":"MD5",
                             "Email_address":"email",
                             "Email1_SHA-1":"SHA1",
                             "MD5 Hashed Email":"MD5",
                             "EE_EMAIL_ADDRESS":"email",
                             "Email1 Address":"email",
                             "DMD_EMAIL":"email",
                             "QBOVRT_master_email":"email",
                             "lower(b.email_addr_hash_txt)":"hashed_email",
                             "Email1_SHA256":"SHA256",
                             "HashedEmail1":"hashed_email",
                             "CONTACT_EMAIL_ADDR":"email",
                             "Email ":"email",
                             "EmailAddressmd5":"MD5",
                             "CONTACT EMAIL":"email",
                             "MD5EMAIL1":"MD5",
                             "MD5email":"MD5",
                             "sha256email\",470121,624331,470121,UPHE - Onboarding - Salesforce Data\"":"SHA256",
                             "sha256email\",287259,579949,28236,Swirl McGarryBowen - Onboarding\"":"SHA256",
                             "AX_CUREMAIL_10":"email",
                             "Email ":"email",
                             "SHA256_email":"SHA256",
                             "USA_EMAIL":"email",
                             "Right_email":"email",
                             "MDG EMAIL":"MD5",
                             "email3":"email",
                             "sha256email\",287259,579949,28236,Swirl McGarryBowen - Onboarding\"":"SHA256",
                             "sha256email\",470121,624331,470121,UPHE - Onboarding - Salesforce Data\"":"SHA256",
                             "email_SHA_256":"SHA256",
                             "hashed_emailId":"hashed_email",
                             "EMAIL_ADDRESS1":"email",
                             "email_adr1":"email",
                             "EmailAddressHash_MD5":"MD5",
                             "LRI_G_email_address":"email",
                             "DANDRVINSON@EMAIL.COM":"email",
                             "user_email":"email",
                             "Email_hashed":"hashed_email",
                             "P_EMAIL":"email",
                             "Hashed email":"hashed_email",
                             "SHA250Email":"SHA256",
                             "sha256email\",287259,579949,28236,Swirl McGarryBowen - Onboarding\"":"SHA256",
                             "sha256email\",470121,624331,470121,UPHE - Onboarding - Salesforce Data\"":"SHA256",
                             "Email5":"email",
                             "Orders Customer Email":"email",
                             "Encrypted Email address":"hashed_email",
                             "eMail":"email",
                             "patron_email_addr":"email",
                             "Client_Email":"email",
                             "email_01":"email",
                             "emailhash_md5":"MD5",
                             "Person Account: Email":"email",
                             "Email_Appended":"email",
                             "Email_address_hash":"hashed_email",
                             "sha256email\",470121,624331,470121,UPHE - Onboarding - Salesforce Data\"":"SHA256",
                             "EMAIL_HASH":"hashed_email",
                             "Email ID":"email",
                             "email_billing":"email",
                             "PRIMARY_EMAIL_ADDRESS":"email",
                             "Opportunity Email":"email",
                             "CnPh_1_01_Email":"email",
                             "MD5_email":"MD5",
                             "email_address_256hash":"SHA256",
                             "EMAIL ADDR":"email",
                             "sha256email\",154996,542386,15314,StubHub - Salesforce Onboarding\"":"SHA256",
                             "Hash Email":"hashed_email",
                             "prmry_email_addr":"email",
                             "sha256email\",248499,508256,24412,Caesars - Salesforce RFP\"":"SHA256",
                             "MEMBER_EMAIL":"email",
                             "Sha1_Email":"SHA1",
                             "HASH EMAIL ADDR":"hashed_email",
                             "Sha1_Email":"SHA1",
                             "HASH EMAIL ADDR":"hashed_email",
                             "CUS_EMAIL_ADDRESS":"email",
                             "channels.email.address":"email",
                             "\Email1\"\"":"email",
                             "primaryEmail":"email",
                             "sha256email\",313789,533849,313789,Audi - Salesforce Imports\"":"SHA256",
                             "HASEHD_EMAIL":"hashed_email",
                             "\Email1\"\"":"email",
                             "hash email":"hashed_email",
                             "REGI_EMAIL_ADDR":"email",
                             "Customers Email":"email",
                             "email_addres":"email",
                             "sha256email\",116909,533639,11536,Weight Watchers - Onboarding\"":"SHA256",
                             "EMAIL-sha1":"SHA1",
                             "Email Address 1":"email",
                             "Contact Business Email":"email",
                             "\Email1\"\"":"email",
                             "Personal_email":"email",
                             "sha256email\",154996,542386,15314,StubHub - Salesforce Onboarding\"":"SHA256",
                             "EMAIL_MATCH_ID_1":"email",
                             "MD5 Email":"MD5",
                             "associatedemailaddress":"email",
                             "Email_Hash":"hashed_email",
                             "Email_Hash ":"hashed_email",
                             "EMAIL ADDRESS	":"email",
                             "EMAILADDR":"email",
                             "EMAILADDR ":"email",
                             "Contact: Primary Email":"email",
                             "Email Address - Prospects":"email",
                             "sha_email	":"hashed_email",
                             "sha_email":"hashed_email",
                             "primary_contact_primary_email":"email",
                             "EMAIL_ADDR_1":"email",
                             "Primary_Email_Address":"email",
                             "CUST_EMAIL_ADDR":"email",
                             "AlternateEmail":"email"
                            }, inplace = True)                       

In [145]:
raw_email_values = pd.DataFrame(raw_emails_all_audiences["label"].value_counts())
raw_email_values.head()
raw_email_aud_counts = raw_email_values["label"][0]
MD5_aud_counts = raw_email_values["label"][1]
SHA256_aud_counts = raw_email_values["label"][3]
SHA1_aud_counts = raw_email_values["label"][4]

query all audiences with AK like '%email%' -> enabled, non-FBR audiences, enabled customers, data store

In [148]:
sql = """
SELECT
  audience_key_labels.id AS ak_id,
  liveramp_customer_accounts.id,
  audience_key_labels.key_index,
  audience_key_labels.label,
  liveramp_customer_accounts.customer_id,
  customers.name,
  liveramp_customer_accounts.display_name,
  COUNT(audience_key_labels.id) AS count_AKs
FROM
  `corp-bi-us-prod.rldb.audience_key_labels` AS audience_key_labels
INNER JOIN
  `corp-bi-us-prod.rldb.liveramp_customer_accounts` AS liveramp_customer_accounts
ON
  audience_key_labels.liveramp_customer_account_id = liveramp_customer_accounts.id
INNER JOIN
  `corp-bi-us-prod.rldb.customers` AS customers
ON
  liveramp_customer_accounts.customer_id = customers.id
WHERE
  liveramp_customer_accounts.id IN (
              SELECT
                lr_customer_accounts.id
              FROM
                `corp-bi-us-prod.rldb.dms_segments` AS dms_segments
              INNER JOIN
                `corp-bi-us-prod.rldb.lrc_field_definitions` AS lrc_field_definitions
              ON
                dms_segments.lrc_field_definition_id = lrc_field_definitions.id
              INNER JOIN
                `corp-bi-us-prod.rldb.liveramp_customer_accounts` AS lr_customer_accounts
              ON
                lrc_field_definitions.lr_customer_account_id = lr_customer_accounts.id
              INNER JOIN
                `corp-bi-us-prod.rldb.customers` AS customers
              ON
                lr_customer_accounts.customer_id = customers.id
              WHERE
                lr_customer_accounts.enabled = TRUE
                AND lr_customer_accounts.customer_link_audience = FALSE
                AND customers.status = 1
              GROUP BY
                lr_customer_accounts.id)
  AND customers.status = 1
  AND liveramp_customer_accounts.enabled = TRUE
  AND liveramp_customer_accounts.customer_link_audience = FALSE
  AND LOWER(audience_key_labels.label) LIKE '%email%'
  AND audience_key_labels.key_index = 0
GROUP BY
  liveramp_customer_accounts.id,
  liveramp_customer_accounts.display_name,
  liveramp_customer_accounts.customer_id,
  customers.name,
  audience_key_labels.key_index,
  ak_id,
  audience_key_labels.label
HAVING
  count_AKs = 1
"""
raw_emails_DS = pandas_gbq.read_gbq(sql, project_id='liveramp-ts-bigquery')

Downloading: 100%|██████████| 112/112 [00:00<00:00, 468.37rows/s]


In [151]:
# normalize email columns (raw, MD5, SHA1, SHA256)
raw_emails_DS["label"].replace({"Email":"email", 
                             "Email1":"email",
                             "Email Address":"email",
                             "EmailAddress":"email", 
                             "Emails":"email",
                             "EMAIL":"email",
                             "Email_1":"email",
                             "EmailAddress":"email",
                             "email_address":"email",
                             "EMAIL_ADDRESS":"email",
                             "Email 1":"email",
                             "Email(s)":"email",
                             "email1":"email",
                             "EMAIL_ID":"email",
                             "email1_MD5":"MD5",
                             "emailmd5":"MD5",
                             "profiles.email_address":"email",
                             "EMAIL_ADDR":"email",
                             "SHA256EMAIL1":"email",
                             "email_addr":"email",
                             "MD5EMAIL":"MD5",
                             "EmailAddress":"email",
                             "EMAIL1":"email",
                             "cust_email_addr":"email",
                             "Email_Address":"email",
                             "md5_email":"MD5",
                             "Email_ID":"email",
                             "EMAIL_ADDR_TXT":"email",
                             "emailAddress":"email",
                             "sha1_email":"SHA1",
                             "MD5Email":"MD5",
                             "email_1":"email",
                             "LRI_G_SHA1EMAIL1":"SHA1",
                             "LRI_G_SHA256EMAIL1":"SHA256",
                             "loweremailmd5":"MD5",
                             "email_sha256":"SHA256",
                             "SHA256_EMAIL":"SHA256",
                             "MD5_Email":"MD5",
                             "LRI_DER_SHA256EMAIL1":"SHA256",
                             "email_SHA-1":"SHA1",
                             "sha256email":"SHA256",
                             "email address":"email",
                             "purch_email_addr":"email",
                             "HashedEmail":"hashed_email",
                             "P_EMAIL_ADDRESS":"email",
                             "Member Email Address":"email",
                             "Email Addresses":"email",
                             "email_md5":"MD5",
                             "emails":"email",
                             "Email Hash":"hashed_email",
                             "customer_email":"email",
                             "EMAIL_ADDRESS_":"email",
                             "SHA 256 Email":"SHA256",
                             "EMAILADDRESS":"email",
                             "LRI_DER_SHA1EMAIL1":"SHA1",
                             "email_hash":"hashed_email",
                             "LRI_G_MD5EMAIL1":"MD5",
                             "EMAIL_SHA256":"SHA256",
                             "PII_EMAIL_ADDRESS":"email",
                             "EMail":"email",
                             "Email address":"email",
                             "SHA1_Email":"SHA1",
                             "LRI_DER_MD5EMAIL1":"MD5",
                             "EMAIL_1":"email",
                             "sha256_email":"SHA256",
                             "SHA256EMAIL":"SHA256",
                             "SHA256Email":"email",
                             "emailaddress":"email",
                             "email_id":"email",
                             "SHA1EMAIL":"SHA1",
                             "hashed email":"hashed_email",
                             "SHA256HASHEDEMAIL":"SHA256",
                             "email2":"email",
                             "EMAIL_ADDRESS_1":"email",
                             "md5email":"MD5",
                             "hashedemail":"hashed_email",
                             "EMAIL_HASHED":"hashed_email",
                             "Broker Email":"email",
                             "HASH_MD5_LOWER_EMAIL":"MD5",
                             "profileEmail":"email",
                             "email_Address":"email",
                             "Hashed_Email":"hashed_email",
                             "PER_EMAIL":"email",
                             "Hashed_Email_Address":"email",
                             "Peloton Order Order Email":"email",
                             "EMAIL,FN,LN,,ZIP,CT,PHONE,":"email",
                             "Hashed_Email_ID_Hex":"hashed_email",
                             "#email_lc_md5":"MD5",
                             "MD5_EMAIL":"MD5",
                             "\EMAIL\"":"email",
                             "Customer Email":"email",
                             "CUSTOMERS_SalesOrder_BILLTOLookup::Email":"email",
                             "profile.email":"email",
                             "HASHED_EMAIL":"hashed_email",
                             "sha256email1":"SHA256",
                             "Hashed Email":"hashed_email",
                             "CUST_EMAIL":"email",
                             "SHA256 Email":"SHA256",
                             "email,fn,ln,st,zip,ct,phone,":"email",
                             "best_email":"email",
                             "LRI_G_EMAIL1":"email",
                             "SHA2_EMAIL":"SHA2",
                             "Md5 Email":"MD5",
                             "md5(email1)":"MD5",
                             "master_email":"email",
                             "HASHED_CUST_EMAIL":"hashed_email",
                             "REAL_EMAIL":"email",
                             "email_sha1":"SHA1",
                             "MD5 EMAIL":"MD5",
                             "Email/Email":"email",
                             "hash_email":"hashed_email",
                             "email md5":"MD5",
                             "Email_address":"email",
                             "Email1_SHA-1":"SHA1",
                             "MD5 Hashed Email":"MD5",
                             "EE_EMAIL_ADDRESS":"email",
                             "Email1 Address":"email",
                             "DMD_EMAIL":"email",
                             "QBOVRT_master_email":"email",
                             "lower(b.email_addr_hash_txt)":"hashed_email",
                             "Email1_SHA256":"SHA256",
                             "HashedEmail1":"hashed_email",
                             "CONTACT_EMAIL_ADDR":"email",
                             "Email ":"email",
                             "EmailAddressmd5":"MD5",
                             "CONTACT EMAIL":"email",
                             "MD5EMAIL1":"MD5",
                             "MD5email":"MD5",
                             "sha256email\",470121,624331,470121,UPHE - Onboarding - Salesforce Data\"":"SHA256",
                             "sha256email\",287259,579949,28236,Swirl McGarryBowen - Onboarding\"":"SHA256",
                             "AX_CUREMAIL_10":"email",
                             "Email ":"email",
                             "SHA256_email":"SHA256",
                             "USA_EMAIL":"email",
                             "Right_email":"email",
                             "MDG EMAIL":"MD5",
                             "email3":"email",
                             "sha256email\",287259,579949,28236,Swirl McGarryBowen - Onboarding\"":"SHA256",
                             "sha256email\",470121,624331,470121,UPHE - Onboarding - Salesforce Data\"":"SHA256",
                             "email_SHA_256":"SHA256",
                             "hashed_emailId":"hashed_email",
                             "EMAIL_ADDRESS1":"email",
                             "email_adr1":"email",
                             "EmailAddressHash_MD5":"MD5",
                             "LRI_G_email_address":"email",
                             "DANDRVINSON@EMAIL.COM":"email",
                             "user_email":"email",
                             "Email_hashed":"hashed_email",
                             "P_EMAIL":"email",
                             "Hashed email":"hashed_email",
                             "SHA250Email":"SHA256",
                             "sha256email\",287259,579949,28236,Swirl McGarryBowen - Onboarding\"":"SHA256",
                             "sha256email\",470121,624331,470121,UPHE - Onboarding - Salesforce Data\"":"SHA256",
                             "Email5":"email",
                             "Orders Customer Email":"email",
                             "Encrypted Email address":"hashed_email",
                             "eMail":"email",
                             "patron_email_addr":"email",
                             "Client_Email":"email",
                             "email_01":"email",
                             "emailhash_md5":"MD5",
                             "Person Account: Email":"email",
                             "Email_Appended":"email",
                             "Email_address_hash":"hashed_email",
                             "sha256email\",470121,624331,470121,UPHE - Onboarding - Salesforce Data\"":"SHA256",
                             "EMAIL_HASH":"hashed_email",
                             "Email ID":"email",
                             "email_billing":"email",
                             "PRIMARY_EMAIL_ADDRESS":"email",
                             "Opportunity Email":"email",
                             "CnPh_1_01_Email":"email",
                             "MD5_email":"MD5",
                             "email_address_256hash":"SHA256",
                             "EMAIL ADDR":"email",
                             "sha256email\",154996,542386,15314,StubHub - Salesforce Onboarding\"":"SHA256",
                             "Hash Email":"hashed_email",
                             "prmry_email_addr":"email",
                             "sha256email\",248499,508256,24412,Caesars - Salesforce RFP\"":"SHA256",
                             "MEMBER_EMAIL":"email",
                             "Sha1_Email":"SHA1",
                             "HASH EMAIL ADDR":"hashed_email",
                             "Sha1_Email":"SHA1",
                             "HASH EMAIL ADDR":"hashed_email",
                             "CUS_EMAIL_ADDRESS":"email",
                             "channels.email.address":"email",
                             "\Email1\"\"":"email",
                             "primaryEmail":"email",
                             "sha256email\",313789,533849,313789,Audi - Salesforce Imports\"":"SHA256",
                             "HASEHD_EMAIL":"hashed_email",
                             "\Email1\"\"":"email",
                             "hash email":"hashed_email",
                             "REGI_EMAIL_ADDR":"email",
                             "Customers Email":"email",
                             "email_addres":"email",
                             "sha256email\",116909,533639,11536,Weight Watchers - Onboarding\"":"SHA256",
                             "EMAIL-sha1":"SHA1",
                             "Email Address 1":"email",
                             "Contact Business Email":"email",
                             "\Email1\"\"":"email",
                             "Personal_email":"email",
                             "sha256email\",154996,542386,15314,StubHub - Salesforce Onboarding\"":"SHA256",
                             "EMAIL_MATCH_ID_1":"email",
                             "MD5 Email":"MD5",
                             "associatedemailaddress":"email",
                             "Email_Hash":"hashed_email",
                             "Email_Hash ":"hashed_email",
                             "EMAIL ADDRESS	":"email",
                             "EMAILADDR":"email",
                             "EMAILADDR ":"email",
                             "Contact: Primary Email":"email",
                             "Email Address - Prospects":"email",
                             "sha_email	":"hashed_email",
                             "sha_email":"hashed_email",
                             "primary_contact_primary_email":"email",
                             "EMAIL_ADDR_1":"email",
                             "Primary_Email_Address":"email",
                             "CUST_EMAIL_ADDR":"email",
                             "AlternateEmail":"email",
                             "PERSONEMAIL":"email",
                             "email_address_01":"email"
                            }, inplace = True)                       

In [157]:
raw_emails_DS_values = pd.DataFrame(raw_emails_DS["label"].value_counts())
raw_emails_DS_values

raw_email_DS_aud_counts = raw_emails_DS_values["label"][0]
MD5_DS_aud_counts = raw_emails_DS_values["label"][1]
SHA256_DS_aud_counts = raw_emails_DS_values["label"][3]
SHA1_DS_aud_counts = raw_emails_DS_values["label"][2]
print("raw email audiences:", raw_email_DS_aud_counts)
print("MD5 email:", MD5_DS_aud_counts)
print("SHA256:", SHA256_DS_aud_counts)
print("SHA1:", SHA1_DS_aud_counts)

raw email audiences: 82
MD5 email: 15
SHA256: 6
SHA1: 8


# CID-only audiences 

In [183]:
## Results
print("Audiences with AK as CID:", CIDs_all_audiences["audience_id"].count())


# offline audiences w/AK as CID - 4,292

# online audiences w/AK as CID - 1,162

# other audiences w/AK as CID - 249

# -------------


# ### non-DS audiences
# enabled, non-FBR audiences where AK = CID and customer = enabled

# non DS audiences w/AK set as CID - 5,285

# non DS offline audiences w/AK as CID - 4050

# non DS online audiences w/AK as CID - 986

# -------------


# ### DS audiences
# enabled, non-FBR audiences where AK = CID and customer = enabled

# DS audiences w/AK set as CID - 418

# DS offline audiences w/AK as CID - 242

# DS online audiences w/AK as CID - 176

# -------------


Audiences with AK as CID: 5786


query all audiences with AK like '%id%' -> enabled, non-FBR audiences, enabled customers, not like: aaid, IDFA, cookie, device id, email, pel, identity_link, idl, liveramp, LR ID 1, LRID, MAID, mobile, order_id, PELID

CID definition methodology: after combing through AK labels strings that match AK like '%id%', IDs that do not resemble CIDs were removed. This analysis thus treats a CID as a synthetic identifier distinguished from IDLs/PELs, cookies, and device IDs.

In [196]:
sql = """
SELECT
  liveramp_customer_accounts.id AS audience_id,
  customers.id AS cust_id, 
  audience_key_labels.key_index,
  lower(audience_key_labels.label) AS aks,
  COUNT(audience_key_labels.id) AS count_aks
FROM
  `corp-bi-us-prod.rldb.audience_key_labels` AS audience_key_labels
INNER JOIN
  `corp-bi-us-prod.rldb.liveramp_customer_accounts` AS liveramp_customer_accounts
ON
  audience_key_labels.liveramp_customer_account_id = liveramp_customer_accounts.id
INNER JOIN
  `corp-bi-us-prod.rldb.customers` AS customers
ON
  liveramp_customer_accounts.customer_id = customers.id
WHERE
  audience_key_labels.key_index = 0
  AND lower(audience_key_labels.label) LIKE '%id%'
  AND lower(audience_key_labels.label) NOT LIKE '%aaid%'
  AND lower(audience_key_labels.label) NOT LIKE '%idfa%'
  AND lower(audience_key_labels.label) NOT LIKE '%cookie%'
  AND lower(audience_key_labels.label) NOT LIKE '%device id%'
  AND lower(audience_key_labels.label) NOT LIKE '%device_id%'
  AND lower(audience_key_labels.label) NOT LIKE '%device%'
  AND lower(audience_key_labels.label) NOT LIKE '%email%'
  AND lower(audience_key_labels.label) NOT LIKE '%pel%'
  AND lower(audience_key_labels.label) NOT LIKE '%identity_link%'
  AND lower(audience_key_labels.label) NOT LIKE '%idl%'
  AND lower(audience_key_labels.label) NOT LIKE '%liveramp%'
  AND lower(audience_key_labels.label) NOT LIKE '%lr id 1%'
  AND lower(audience_key_labels.label) NOT LIKE '%lrid%'
  AND lower(audience_key_labels.label) NOT LIKE '%maid%'
  AND lower(audience_key_labels.label) NOT LIKE '%mobile%'
  AND lower(audience_key_labels.label) NOT LIKE '%order_id%'
  AND lower(audience_key_labels.label) NOT LIKE '%pelid%'
  AND lower(audience_key_labels.label) NOT LIKE '%fb_ext_id%'
  AND lower(audience_key_labels.label) NOT LIKE '%identitylink%'
  AND lower(audience_key_labels.label) NOT LIKE '%muid%'
  AND lower(audience_key_labels.label) NOT LIKE '%abilitec%'
  AND lower(audience_key_labels.label) NOT LIKE '%acxiom_id%'
  AND lower(audience_key_labels.label) NOT LIKE '%extern_id%'
  AND lower(audience_key_labels.label) NOT LIKE '%cnsmr_link_id%'
  AND lower(audience_key_labels.label) NOT LIKE '%ttd_android%'
  AND customers.status = 1
  AND liveramp_customer_accounts.enabled = TRUE
  AND liveramp_customer_accounts.customer_link_audience = FALSE
  AND audience_key_labels.key_index = 0
GROUP BY
  1, 2, 3, 4
HAVING
  COUNT(audience_key_labels.id) = 1
"""

CIDs_all_audiences = pandas_gbq.read_gbq(sql, project_id='liveramp-ts-bigquery')
CIDs_all_audiences

Downloading: 100%|██████████| 5786/5786 [00:00<00:00, 9302.75rows/s]


audience_id  cust_id  key_index           aks  count_aks
0          501221   550486          0    customerid          1
1          502581   586779          0           sid          1
2          247989   554196          0            id          1
3          246779   540626          0            id          1
4          246629   506766          0           cid          1
...           ...      ...        ...           ...        ...
5781        43046   394608          0        dfa_id          1
5782        43646   514366          0        userid          1
5783        43856   405604          0       cons_id          1
5784        42746   512256          0  voterbase_id          1
5785         9898   487118          0     dh_hh_id          1

[5786 rows x 5 columns]

In [ ]:
sql = """
SELECT
  liveramp_customer_accounts.id AS audience_id,
  customers.id AS cust_id, 
  audience_key_labels.key_index,
  lower(audience_key_labels.label) AS aks,
  COUNT(audience_key_labels.id) AS count_aks
FROM
  `corp-bi-us-prod.rldb.audience_key_labels` AS audience_key_labels
INNER JOIN
  `corp-bi-us-prod.rldb.liveramp_customer_accounts` AS liveramp_customer_accounts
ON
  audience_key_labels.liveramp_customer_account_id = liveramp_customer_accounts.id
INNER JOIN
  `corp-bi-us-prod.rldb.customers` AS customers
ON
  liveramp_customer_accounts.customer_id = customers.id
WHERE
  audience_key_labels.key_index = 0
  AND lower(audience_key_labels.label) LIKE '%id%'
  AND lower(audience_key_labels.label) NOT LIKE '%aaid%'
  AND lower(audience_key_labels.label) NOT LIKE '%idfa%'
  AND lower(audience_key_labels.label) NOT LIKE '%cookie%'
  AND lower(audience_key_labels.label) NOT LIKE '%device id%'
  AND lower(audience_key_labels.label) NOT LIKE '%device_id%'
  AND lower(audience_key_labels.label) NOT LIKE '%device%'
  AND lower(audience_key_labels.label) NOT LIKE '%email%'
  AND lower(audience_key_labels.label) NOT LIKE '%pel%'
  AND lower(audience_key_labels.label) NOT LIKE '%identity_link%'
  AND lower(audience_key_labels.label) NOT LIKE '%idl%'
  AND lower(audience_key_labels.label) NOT LIKE '%liveramp%'
  AND lower(audience_key_labels.label) NOT LIKE '%lr id 1%'
  AND lower(audience_key_labels.label) NOT LIKE '%lrid%'
  AND lower(audience_key_labels.label) NOT LIKE '%maid%'
  AND lower(audience_key_labels.label) NOT LIKE '%mobile%'
  AND lower(audience_key_labels.label) NOT LIKE '%order_id%'
  AND lower(audience_key_labels.label) NOT LIKE '%pelid%'
  AND lower(audience_key_labels.label) NOT LIKE '%fb_ext_id%'
  AND lower(audience_key_labels.label) NOT LIKE '%identitylink%'
  AND lower(audience_key_labels.label) NOT LIKE '%muid%'
  AND lower(audience_key_labels.label) NOT LIKE '%abilitec%'
  AND lower(audience_key_labels.label) NOT LIKE '%acxiom_id%'
  AND lower(audience_key_labels.label) NOT LIKE '%extern_id%'
  AND lower(audience_key_labels.label) NOT LIKE '%cnsmr_link_id%'
  AND lower(audience_key_labels.label) NOT LIKE '%ttd_android%'
  AND customers.status = 1
  AND liveramp_customer_accounts.enabled = TRUE
  AND liveramp_customer_accounts.customer_link_audience = FALSE
  AND audience_key_labels.key_index = 0
GROUP BY
  1, 2, 3, 4
HAVING
  COUNT(audience_key_labels.id) = 1
"""

CIDs_all_audiences = pandas_gbq.read_gbq(sql, project_id='liveramp-ts-bigquery')
CIDs_all_audiences

In [ ]:
# CIDs_all_audiences = pd.read_csv('/Users/aschul/Documents/Documents - aschul1118mac/Code_Base/Investigations/20200417_AK_analysis/CID_AKs/query_result.csv')

In [ ]:
CIDs_all_audiences.head()

In [179]:
CID_values = pd.DataFrame(CIDs_all_audiences["aks"].value_counts())

In [180]:
CID_values.head(10)

aks
customer id         608
client customer id  559
customerid          390
id                  345
customer_id         310
ccid                270
record_id           135
mbid                126
cid                 103
client_customer_id   92

In [184]:
CID_values.info()

<class 'pandas.core.frame.DataFrame'>
Index: 672 entries, customer id to css_client_id
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   aks     672 non-null    int64
dtypes: int64(1)
memory usage: 10.5+ KB


In [185]:
CID_values["aks"]

customer id              608
client customer id       559
customerid               390
id                       345
customer_id              310
                        ... 
vb_voterbase_id            1
audience_idnt              1
epsilon individual id      1
hashed-salted-devid        1
css_client_id              1
Name: aks, Length: 672, dtype: int64

query all audiences with AK like '%ID%' -> enabled, non-FBR audiences, enabled customers, data store

In [ ]:
sql = """
SELECT
  liveramp_customer_accounts.id AS audience_id,
  customers.id AS cust_id, 
  audience_key_labels.key_index AS key_index,
  lower(audience_key_labels.label) AS label,
  COUNT(audience_key_labels.id) AS count_AK
FROM
  `corp-bi-us-prod.rldb.audience_key_labels` AS audience_key_labels
INNER JOIN
  `corp-bi-us-prod.rldb.liveramp_customer_accounts` AS liveramp_customer_accounts
ON
  audience_key_labels.liveramp_customer_account_id = liveramp_customer_accounts.id
INNER JOIN
  `corp-bi-us-prod.rldb.customers` AS customers
ON
  liveramp_customer_accounts.customer_id = customers.id
WHERE
liveramp_customer_accounts.id IN (
              SELECT
                lr_customer_accounts.id
              FROM
                `corp-bi-us-prod.rldb.dms_segments` AS dms_segments
              INNER JOIN
                `corp-bi-us-prod.rldb.lrc_field_definitions` AS lrc_field_definitions
              ON
                dms_segments.lrc_field_definition_id = lrc_field_definitions.id
              INNER JOIN
                `corp-bi-us-prod.rldb.liveramp_customer_accounts` AS lr_customer_accounts
              ON
                lrc_field_definitions.lr_customer_account_id = lr_customer_accounts.id
              INNER JOIN
                `corp-bi-us-prod.rldb.customers` AS customers
              ON
                lr_customer_accounts.customer_id = customers.id
              WHERE
                lr_customer_accounts.enabled = TRUE
                AND lr_customer_accounts.customer_link_audience = FALSE
                AND customers.status = 1
              GROUP BY
                lr_customer_accounts.id) 
  AND 
  audience_key_labels.key_index = 0
  AND lower(audience_key_labels.label) LIKE '%id%'
  AND lower(audience_key_labels.label) NOT LIKE '%aaid%'
  AND lower(audience_key_labels.label) NOT LIKE '%idfa%'
  AND lower(audience_key_labels.label) NOT LIKE '%cookie%'
  AND lower(audience_key_labels.label) NOT LIKE '%device id%'
  AND lower(audience_key_labels.label) NOT LIKE '%device_id%'
  AND lower(audience_key_labels.label) NOT LIKE '%device%'
  AND lower(audience_key_labels.label) NOT LIKE '%email%'
  AND lower(audience_key_labels.label) NOT LIKE '%pel%'
  AND lower(audience_key_labels.label) NOT LIKE '%identity_link%'
  AND lower(audience_key_labels.label) NOT LIKE '%idl%'
  AND lower(audience_key_labels.label) NOT LIKE '%liveramp%'
  AND lower(audience_key_labels.label) NOT LIKE '%lr id 1%'
  AND lower(audience_key_labels.label) NOT LIKE '%lrid%'
  AND lower(audience_key_labels.label) NOT LIKE '%maid%'
  AND lower(audience_key_labels.label) NOT LIKE '%mobile%'
  AND lower(audience_key_labels.label) NOT LIKE '%order_id%'
  AND lower(audience_key_labels.label) NOT LIKE '%pelid%'
  AND lower(audience_key_labels.label) NOT LIKE '%fb_ext_id%'
  AND lower(audience_key_labels.label) NOT LIKE '%identitylink%'
  AND lower(audience_key_labels.label) NOT LIKE '%muid%'
  AND lower(audience_key_labels.label) NOT LIKE '%abilitec%'
  AND lower(audience_key_labels.label) NOT LIKE '%acxiom_id%'
  AND lower(audience_key_labels.label) NOT LIKE '%extern_id%'
  AND lower(audience_key_labels.label) NOT LIKE '%cnsmr_link_id%'
  AND lower(audience_key_labels.label) NOT LIKE '%ttd_android%'
  AND customers.status = 1
  AND liveramp_customer_accounts.enabled = TRUE
  AND liveramp_customer_accounts.customer_link_audience = FALSE
  AND audience_key_labels.key_index = 0
  AND liveramp_customer_accounts.type = "OfflineLiverampCustomerAccount"
GROUP BY
  1, 2, 3, 4
HAVING
  COUNT(audience_key_labels.id) = 1
"""
CIDs_DS_audiences = pandas_gbq.read_gbq(sql, project_id='liveramp-ts-bigquery')


In [ ]:
# CIDs_DS_audiences = pd.read_csv('/Users/aschul/Documents/Documents - aschul1118mac/Code_Base/Investigations/20200417_AK_analysis/CID_AKs/query_result_DS.csv')

In [ ]:
CIDs_DS_audiences.head(5)

In [ ]:
CID_values_DS_audiences = pd.DataFrame(CIDs_DS_audiences["label"].value_counts())

In [ ]:
CID_values_DS_audiences.head(10)

# Address-only audiences


## Results

### all audiences
enabled, non-FBR audiences where customer = enabled and AK = address

all audiences w/ak = address -> 1,145

offline audiences w/ak = address -> 1,133

online audiences w/ak = address -> 12

### data store

all audiences w/ak = address -> 90

offline audiences w/ak = address -> 90

online audiences w/ak = address -> 0


## Data Wrangling

query all audiences for # of aks between 4 and 7 inclusive, enabled customers, enabled audiences, non-FBR

In [198]:
sql = """
SELECT
  audience_key_labels.liveramp_customer_account_id,
  customers.id AS customer_id,
  customers.name AS customer_name,
  COUNT(audience_key_labels.id) AS ak_count,
  STRING_AGG(LOWER(audience_key_labels.label)) AS aks,
  liveramp_customer_accounts.type
FROM
  `corp-bi-us-prod.rldb.audience_key_labels` AS audience_key_labels
INNER JOIN
  `corp-bi-us-prod.rldb.liveramp_customer_accounts` AS liveramp_customer_accounts
ON
  audience_key_labels.liveramp_customer_account_id = liveramp_customer_accounts.id
INNER JOIN
  `corp-bi-us-prod.rldb.customers` AS customers
ON
  liveramp_customer_accounts.customer_id = customers.id
WHERE
  customers.status = 1
  AND liveramp_customer_accounts.enabled = TRUE
  AND liveramp_customer_accounts.customer_link_audience = FALSE

GROUP BY
  1,
  2,
  3,
  6
HAVING
  ak_count > 3
  AND ak_count < 8
  AND aks NOT LIKE "%transaction%"
  AND aks NOT LIKE "%amount%"
  AND aks NOT LIKE "%aaid%"
  AND aks NOT LIKE "%email%"
  AND aks NOT LIKE "%ordinal%"
ORDER BY
  ak_count DESC
"""
address_aks = pandas_gbq.read_gbq(sql, project_id='liveramp-ts-bigquery')

Downloading: 100%|██████████| 1162/1162 [00:00<00:00, 2901.21rows/s]


In [199]:
address_aks.head(5)

liveramp_customer_account_id  customer_id            customer_name  \
0                        503441       517576             Arena Online   
1                        248439       517576             Arena Online   
2                        217766       517576             Arena Online   
3                        217736       517576             Arena Online   
4                        216436       526846  Social Reality - SRAXmd   

   ak_count                                                aks  \
0         7  customer id,last name,city,address ,state,firs...   
1         7  city,state,zip code,last name,first name,custo...   
2         7  city,state,zip code,last name,first name,stree...   
3         7  city,state,zip code,last name,first name,stree...   
4         7  city,state,address1,address2,zip code,last nam...   

                             type  
0  OfflineLiverampCustomerAccount  
1  OfflineLiverampCustomerAccount  
2  OfflineLiverampCustomerAccount  
3  OfflineLiverampCustomerAccount  
4  OfflineLiverampCustomerAccount

In [ ]:
# count types

query all audiences for # of aks between 4 and 7 inclusive, enabled customers, enabled audiences, non-FBR, DS

In [200]:
sql = """
SELECT
  audience_key_labels.liveramp_customer_account_id,
  customers.id AS customer_id,
  customers.name AS customer_name,
  COUNT(audience_key_labels.id) AS ak_count,
  STRING_AGG(LOWER(audience_key_labels.label)) AS aks,
  liveramp_customer_accounts.type
FROM
  `corp-bi-us-prod.rldb.audience_key_labels` AS audience_key_labels
INNER JOIN
  `corp-bi-us-prod.rldb.liveramp_customer_accounts` AS liveramp_customer_accounts
ON
  audience_key_labels.liveramp_customer_account_id = liveramp_customer_accounts.id
INNER JOIN
  `corp-bi-us-prod.rldb.customers` AS customers
ON
  liveramp_customer_accounts.customer_id = customers.id
WHERE
liveramp_customer_accounts.id IN (
              SELECT
                lr_customer_accounts.id
              FROM
                `corp-bi-us-prod.rldb.dms_segments` AS dms_segments
              INNER JOIN
                `corp-bi-us-prod.rldb.lrc_field_definitions` AS lrc_field_definitions
              ON
                dms_segments.lrc_field_definition_id = lrc_field_definitions.id
              INNER JOIN
                `corp-bi-us-prod.rldb.liveramp_customer_accounts` AS lr_customer_accounts
              ON
                lrc_field_definitions.lr_customer_account_id = lr_customer_accounts.id
              INNER JOIN
                `corp-bi-us-prod.rldb.customers` AS customers
              ON
                lr_customer_accounts.customer_id = customers.id
              WHERE
                lr_customer_accounts.enabled = TRUE
                AND lr_customer_accounts.customer_link_audience = FALSE
                AND customers.status = 1
              GROUP BY
                lr_customer_accounts.id) AND
  customers.status = 1
  AND liveramp_customer_accounts.enabled = TRUE
  AND liveramp_customer_accounts.customer_link_audience = FALSE
  AND liveramp_customer_accounts.type = "OfflineLiverampCustomerAccount"
GROUP BY
  1,
  2,
  3,
  6
HAVING
  ak_count > 3
  AND ak_count < 8
  AND aks NOT LIKE "%transaction%"
  AND aks NOT LIKE "%amount%"
  AND aks NOT LIKE "%aaid%"
  AND aks NOT LIKE "%email%"
  AND aks NOT LIKE "%ordinal%"
ORDER BY
  ak_count DESC
"""
address_aks_DS = pandas_gbq.read_gbq(sql, project_id='liveramp-ts-bigquery')

Downloading: 100%|██████████| 90/90 [00:00<00:00, 312.83rows/s]


In [201]:
address_aks_DS

liveramp_customer_account_id  customer_id  \
0                         113519       576169   
1                         177096       437078   
2                          98699       576169   
3                          97599       576169   
4                         341589       576169   
5                         408301       538859   
6                         356929       437078   
7                         134309       144191   
8                         133059       520876   
9                         105779       144191   
10                        483981       437078   
11                        484431       437078   
12                        164476       437078   
13                        164496       437078   
14                        164486       437078   
15                        164466       437078   
16                        164506       437078   
17                        164516       437078   
18                        159466       576169   
19                           617       576169   
20                        206896       521086   
21                        297579       437078   
22                        163946       437078   
23                        382189       576169   
24                        354429       576169   
25                        352589       437078   
26                        352599       437078   
27                        438361       531379   
28                        438351       531379   
29                        400001       437078   
30                         43346       144191   
31                        246559       144191   
32                         98099       576169   
33                         71756       517946   
34                         47836       436078   
35                        436001       538859   
36                        377749       517946   
37                        378609       517946   
38                        378899        57384   
39                        310859       517946   
40                         88826       521086   
41                         55256       144191   
42                        179186        57384   
43                        149996       144191   
44                         70326       517946   
45                        453021       517946   
46                        360659       517946   
47                        360969       517946   
48                        335419       592649   
49                        276419       517946   
50                        212576       517946   
51                        210446       521086   
52                        177856       517946   
53                        408291       538859   
54                        408321       538859   
55                        408331       538859   
56                        408341       538859   
57                        408281       538859   
58                        408311       538859   
59                        408271       538859   
60                        110309       520876   
61                         69846       517946   
62                        388459       517946   
63                        326349       517946   
64                        167686       544656   
65                         80516        57384   
66                         22356       400648   
67                         22876       437078   
68                         13669       213651   
69                        257969       144191   
70                        130109       144191   
71                        129689       509776   
72                         32246       511386   
73                        201756       507516   
74                        405041       517946   
75                        354059       517946   
76                        127729       144191   
77                        352289       560826   
78                        250369       144191   
79                         29536       437078   
80                         17056   

# PINless imports analysis

## results
non DS audiences 
email - 7
CID - 83
address - 2

DS audiences - 103
address - 6
raw email - 8
MD5 - 4
SHA256 - 1
CID - 76

In [ ]:
sql = """
WITH
  pinless_audiences AS 

  (
  SELECT
    liveramp_customer_accounts.display_name AS audience_name,
    liveramp_customer_accounts.id AS audience_id,
    COUNT(liveramp_import_requests.id) AS total_imports,
    COUNT(lir_pin_types.id) AS PIN_imports,
    SUM(CASE
        WHEN lir_pin_types.id IS NULL THEN 1
      ELSE
      0
    END
      ) AS PINLESS_imports,
    SAFE_DIVIDE(SUM(CASE
          WHEN lir_pin_types.id IS NULL THEN 1
        ELSE
        0
      END
        ),
      COUNT(liveramp_import_requests.id)) AS percent_pinless
  FROM
    `corp-bi-us-prod.rldb.liveramp_customer_accounts` AS liveramp_customer_accounts
  INNER JOIN
    `corp-bi-us-prod.rldb.customers` AS customers
  ON
    liveramp_customer_accounts.customer_id = customers.id
  INNER JOIN
    `corp-bi-us-prod.rldb.liveramp_import_requests` AS liveramp_import_requests
  ON
    liveramp_customer_accounts.id = liveramp_import_requests.liveramp_customer_account_id
  LEFT JOIN
    `corp-bi-us-prod.rldb.lir_pin_types` AS lir_pin_types
  ON
    liveramp_import_requests.id = lir_pin_types.liveramp_import_request_id
  WHERE
    liveramp_customer_accounts.type = "OfflineLiverampCustomerAccount"
    AND liveramp_customer_accounts.customer_link_audience = FALSE
    AND liveramp_import_requests.status <> 11
  GROUP BY
    1,
    2
  HAVING
    percent_pinless > 0 
    )

  ,
  audience_aks AS 
  (
  SELECT
    audience_key_labels.liveramp_customer_account_id AS table_2_audience_id,
    customers.id AS customer_id,
    customers.name AS customer_name,
    COUNT(audience_key_labels.id) AS ak_count,
    STRING_AGG(LOWER(audience_key_labels.label)) AS aks,
    liveramp_customer_accounts.type
  FROM
    `corp-bi-us-prod.rldb.audience_key_labels` AS audience_key_labels
  INNER JOIN
    `corp-bi-us-prod.rldb.liveramp_customer_accounts` AS liveramp_customer_accounts
  ON
    audience_key_labels.liveramp_customer_account_id = liveramp_customer_accounts.id
  INNER JOIN
    `corp-bi-us-prod.rldb.customers` AS customers
  ON
    liveramp_customer_accounts.customer_id = customers.id
  WHERE
    customers.status = 1
    AND liveramp_customer_accounts.enabled = TRUE
    AND liveramp_customer_accounts.customer_link_audience = FALSE
  GROUP BY
    1,
    2,
    3,
    6
  ORDER BY
    ak_count DESC
    )

SELECT
  audience_id, audience_name, customer_name, customer_id, total_imports, PIN_imports, PINLESS_imports, percent_pinless, ak_count, aks
FROM
  pinless_audiences
INNER JOIN
  audience_aks
ON
  audience_id = table_2_audience_id
"""

pinless_auds = pandas_gbq.read_gbq(sql, project_id='liveramp-ts-bigquery')
pinless_auds.head()

# dd/mm/YY H:M:S
dt_string = now.strftime("%d/%m/%Y %H:%M:%S")
print("Time of data pull =", dt_string)

In [ ]:
total_auds = pinless_auds.loc[pinless_auds['percent_pinless']>0.198]
total_auds = total_auds.loc[total_auds['total_imports']>31]
total_auds.info()

In [ ]:
# filter for audiences w/one audience key
pinless_auds_1_ak = pinless_auds.loc[pinless_auds['ak_count'] == 1]
# filter for audiences with %pinless imports in 50th percentile
pinless_auds_1_ak_50 = pinless_auds_1_ak.loc[pinless_auds_1_ak['percent_pinless']>0.198]
# 50th percentile number of imports, 50th percentile percent pinless
pinless_auds_1_ak_50_50 = pinless_auds_1_ak_50.loc[pinless_auds_1_ak_50['total_imports']>31]

In [ ]:
pinless_auds_1_ak_50_50 
pinless_auds_1_ak_50_50_values = pd.DataFrame(pinless_auds_1_ak_50_50["aks"].value_counts())

In [ ]:
# audiences w/4-8 aks (addresses)
# filter for audiences w/4-8 audience keys
pinless_auds_4_ak = pinless_auds[(pinless_auds['ak_count'] >= 4) & (pinless_auds['ak_count'] <= 8)]
# filter for audiences with %pinless imports in 50th percentile
pinless_auds_4_ak_50 = pinless_auds_4_ak.loc[pinless_auds_4_ak['percent_pinless']>0.198]
# 50th percentile number of imports, 50th percentile percent pinless
pinless_auds_4_ak_50_50 = pinless_auds_4_ak_50.loc[pinless_auds_4_ak_50['total_imports']>31]

Data Store audiences, query PINless audiences

In [ ]:
sql = """
WITH
  pinless_audiences AS 

  (
  SELECT
    liveramp_customer_accounts.display_name AS audience_name,
    liveramp_customer_accounts.id AS audience_id,
    COUNT(liveramp_import_requests.id) AS total_imports,
    COUNT(lir_pin_types.id) AS PIN_imports,
    SUM(CASE
        WHEN lir_pin_types.id IS NULL THEN 1
      ELSE
      0
    END
      ) AS PINLESS_imports,
    SAFE_DIVIDE(SUM(CASE
          WHEN lir_pin_types.id IS NULL THEN 1
        ELSE
        0
      END
        ),
      COUNT(liveramp_import_requests.id)) AS percent_pinless
  FROM
    `corp-bi-us-prod.rldb.liveramp_customer_accounts` AS liveramp_customer_accounts
  INNER JOIN
    `corp-bi-us-prod.rldb.customers` AS customers
  ON
    liveramp_customer_accounts.customer_id = customers.id
  INNER JOIN
    `corp-bi-us-prod.rldb.liveramp_import_requests` AS liveramp_import_requests
  ON
    liveramp_customer_accounts.id = liveramp_import_requests.liveramp_customer_account_id
  LEFT JOIN
    `corp-bi-us-prod.rldb.lir_pin_types` AS lir_pin_types
  ON
    liveramp_import_requests.id = lir_pin_types.liveramp_import_request_id
  WHERE
    liveramp_customer_accounts.type = "OfflineLiverampCustomerAccount"
    AND liveramp_customer_accounts.customer_link_audience = FALSE
    AND liveramp_import_requests.status <> 11
  GROUP BY
    1,
    2
  HAVING
    percent_pinless > 0 
    )

  ,
  audience_aks AS 
  (
  SELECT
    audience_key_labels.liveramp_customer_account_id AS table_2_audience_id,
    customers.id AS customer_id,
    customers.name AS customer_name,
    COUNT(audience_key_labels.id) AS ak_count,
    STRING_AGG(LOWER(audience_key_labels.label)) AS aks,
    liveramp_customer_accounts.type
  FROM
    `corp-bi-us-prod.rldb.audience_key_labels` AS audience_key_labels
  INNER JOIN
    `corp-bi-us-prod.rldb.liveramp_customer_accounts` AS liveramp_customer_accounts
  ON
    audience_key_labels.liveramp_customer_account_id = liveramp_customer_accounts.id
  INNER JOIN
    `corp-bi-us-prod.rldb.customers` AS customers
  ON
    liveramp_customer_accounts.customer_id = customers.id
  WHERE
    customers.status = 1
    AND liveramp_customer_accounts.enabled = TRUE
    AND liveramp_customer_accounts.customer_link_audience = FALSE
  GROUP BY
    1,
    2,
    3,
    6
  ORDER BY
    ak_count DESC
    )

SELECT
  audience_id, audience_name, customer_name, customer_id, total_imports, PIN_imports, PINLESS_imports, percent_pinless, ak_count, aks
FROM
  pinless_audiences
INNER JOIN
  audience_aks
ON
  audience_id = table_2_audience_id
WHERE audience_id IN (
              SELECT
                lr_customer_accounts.id
              FROM
                `corp-bi-us-prod.rldb.dms_segments` AS dms_segments
              INNER JOIN
                `corp-bi-us-prod.rldb.lrc_field_definitions` AS lrc_field_definitions
              ON
                dms_segments.lrc_field_definition_id = lrc_field_definitions.id
              INNER JOIN
                `corp-bi-us-prod.rldb.liveramp_customer_accounts` AS lr_customer_accounts
              ON
                lrc_field_definitions.lr_customer_account_id = lr_customer_accounts.id
              INNER JOIN
                `corp-bi-us-prod.rldb.customers` AS customers
              ON
                lr_customer_accounts.customer_id = customers.id
              WHERE
                lr_customer_accounts.enabled = TRUE
                AND lr_customer_accounts.customer_link_audience = FALSE
                AND customers.status = 1
              GROUP BY
                lr_customer_accounts.id) 
"""
pinless_auds_ds = pandas_gbq.read_gbq(sql, project_id='liveramp-ts-bigquery')

# dd/mm/YY H:M:S
dt_string = now.strftime("%d/%m/%Y %H:%M:%S")
print("Time of data pull =", dt_string)

In [ ]:
# filter for audiences w/one audience key 
pinless_auds_ds_1 = pinless_auds_ds.loc[pinless_auds_ds['ak_count'] == 1] 
# filter for percent pinless > 0.198
pinless_auds_ds_50 = pinless_auds_ds_1.loc[pinless_auds_ds_1['percent_pinless']>0.198]
# filter for total imports > 32
pinless_auds_ds_50_50 = pinless_auds_ds_50.loc[pinless_auds_ds_50['total_imports']>32]

In [ ]:
# filter for audiences w/4-8 audience keys
pinless_auds_ds_4 = pinless_auds_ds[(pinless_auds_ds['ak_count'] >= 4) & (pinless_auds_ds['ak_count'] <= 8)]
# filter for percent pinless > 0.198
# NO RESULTS

## appendix

In [ ]:
pinless_auds.describe()

In [ ]:
percent_pinless = pinless_auds['percent_pinless']

In [ ]:
percent_pinless.plot.hist()